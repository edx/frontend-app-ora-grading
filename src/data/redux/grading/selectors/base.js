import { createSelector } from 'reselect';

import { StrictDict } from 'utils';
import * as module from './base';

export const rootSelector = ({ grading }) => grading;
export const mkSimpleSelector = (key) => createSelector(
  [module.rootSelector],
  (grading) => grading[key],
);
export const rootKeys = StrictDict({
  selected: 'selected',
  activeIndex: 'activeIndex',
  current: 'current',
  gradeData: 'gradeData',
  gradingData: 'gradingData',
});

export const simpleSelectors = {
  selected: mkSimpleSelector(rootKeys.selected),
  activeIndex: mkSimpleSelector(rootKeys.activeIndex),
  current: mkSimpleSelector(rootKeys.current),
  gradeData: mkSimpleSelector(rootKeys.gradeData),
  gradingData: mkSimpleSelector(rootKeys.gradingData),
};

/**
 * returns the length of the list of selected submissions
 * @return {number} selected submission list length
 */
export const selectionLength = createSelector(
  [module.simpleSelectors.selected],
  (selected) => selected.length,
);

export default {
  ...simpleSelectors,
  selectionLength,
};
