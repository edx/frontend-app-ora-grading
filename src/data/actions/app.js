import { StrictDict } from 'utils';
// import { createActionFactory } from './utils';

// export const dataKey = 'app';
// const createAction = createActionFactory(dataKey);

export default StrictDict({ });
