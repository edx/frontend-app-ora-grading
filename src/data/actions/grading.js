import { StrictDict } from 'utils';
import { createActionFactory } from './utils';

export const dataKey = 'grading';
const createAction = createActionFactory(dataKey);

/**
 * Load the first of the selected submission list for review, and initializes
 * the review pane to the first index.
 * @param {obj} submission data for the review/grading view
 *   {
 *     {obj} response - api response data
 *     {obj} gradeData - api grade data
 *     {str} status - api grade status
 *   }
 */
const loadSubmission = createAction('loadSubmission');

/**
 * Pre-load just the static info about the "next" submission in the review queue.
 * Load submission and the learner's response.
 * @param {obj} submission ({ response })
 */
const preloadNext = createAction('preloadNext');

/**
 * Pre-load just the static info about the "previous" submission in the review queue.
 * Load submission and the learner's response.
 * @param {obj} submission ({ response })
 */
const preloadPrev = createAction('preloadPrev');

/**
 * Load the "next" submission in the selected queue as the current selection, load its current
 * status and grade data, and update prev/next accordingly.
 * @param {obj} { status, gradeData }
 */
const loadNext = createAction('loadNext');

/**
 * Load the "prev" submission in the selected queue as the current selection, load its current
 * status and grade data, and update prev/next accordingly.
 * @param {obj} { status, gradeData }
 */
const loadPrev = createAction('loadPrev');

/**
 * Load the selected submissions, storing their static data in an ordered array and setting the starting
 * index at the beginning of the list.
 * @param {obj[]} selection - ordered array of submission static data for all selected submissions
 */
const updateSelection = createAction('updateSelection');

// TODO: implement/design data workflow
const rubric = StrictDict({
  /*
   * update the local version of the rubric-level comment
   * @param {string} comment
   */
  updateComment: createAction('rubric/comment'),
  /*
   * update the local version of points for the given criterion
   * @param {number} index
   * @param {number} points
   */
  updateCriterionPoints: createAction('rubric/criterionPoints'),
  /*
   * update the local version of comment for the given criterion
   * @param {number} index
   * @param {string} comments
   */
  updateCriterionComment: createAction('rubric/criterionComment'),
});

export default StrictDict({
  loadSubmission,
  preloadNext,
  preloadPrev,
  loadNext,
  loadPrev,
  updateSelection,
  rubric,
});
