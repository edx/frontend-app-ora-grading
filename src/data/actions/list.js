import { StrictDict } from 'utils';
import { createActionFactory } from './utils';

export const dataKey = 'list';
const createAction = createActionFactory(dataKey);

export default StrictDict({
  load: createAction('load'),
  updateSelection: createAction('updateSelection'),
});
