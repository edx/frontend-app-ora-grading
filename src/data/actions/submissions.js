import { StrictDict } from 'utils';
import { createActionFactory } from './utils';

export const dataKey = 'submissions';
const createAction = createActionFactory(dataKey);

/**
 * Load the basic list-level submission data, keyed by submission id
 * @param {obj} submissionListData
 */
const loadList = createAction('loadList');

export default StrictDict({
  loadList,
});
