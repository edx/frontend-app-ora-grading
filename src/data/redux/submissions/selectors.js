import _ from 'lodash';
import { createSelector } from 'reselect';

import { StrictDict } from 'utils';
import { lockStatuses } from 'data/services/lms/constants';

export const simpleSelectors = {
  allSubmissions: state => state.submissions.allSubmissions,
};

/**
 * Returns the submission list in default order for the table.
 */
export const listData = createSelector(
  [simpleSelectors.allSubmissions],
  (allSubmissions) => {
    const submissionIds = Object.keys(allSubmissions);
    const submissionList = submissionIds.map(id => {
      const { gradeStatus, lockStatus, ...rest } = allSubmissions[id];
      const gradingStatus = (lockStatus === lockStatuses.unlocked ? gradeStatus : lockStatus);
      return { gradingStatus, ...rest };
    });
    return _.sortBy(
      submissionList,
      ['submissionDate'],
    );
  },
);

export default StrictDict({
  ...simpleSelectors,
  listData,
});
