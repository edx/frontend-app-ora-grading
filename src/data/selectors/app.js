import { StrictDict } from 'utils';

export const simpleSelectors = {
  showReview: state => state.app.showReview,
  showRubric: state => state.app.showRubric,
  grading: state => state.app.grading,
  courseMetadata: state => state.app.courseMetadata,
  oraName: state => state.app.oraMetadata.name,
  oraPrompt: state => state.app.oraMetadata.prompt,
  oraTypes: state => state.app.oraMetadata.type,
};

export default StrictDict({
  ...simpleSelectors,
});
