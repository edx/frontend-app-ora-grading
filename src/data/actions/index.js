import { StrictDict } from 'utils';

import app from './app';
import grading from './grading';
import submissions from './submissions';

export default StrictDict({
  app,
  grading,
  submissions,
});
