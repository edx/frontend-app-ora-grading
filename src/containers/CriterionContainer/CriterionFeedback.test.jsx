import React from 'react';
import { shallow } from 'enzyme';

import actions from 'data/actions';
import selectors from 'data/selectors';
import {
  CriterionFeedback,
  mapStateToProps,
  mapDispatchToProps,
} from './CriterionFeedback';

jest.mock('@edx/paragon', () => ({
  Form: {
    Control: () => 'Form.Control',
  },
}));

jest.mock('data/selectors', () => ({
  __esModule: true,
  default: {
    app: {
      rubric: {
        criterionFeedbackConfig: jest.fn((...args) => ({
          _rubricCriterionFeedbackConfig: args,
        })),
      },
    },
    grading: {
      selected: {
        criterionFeedback: jest.fn((...args) => ({ _criterionFeedback: args })),
        gradeStatus: jest.fn((...args) => ({ _gradeStatus: args })),
      },
    },
  },
}));

describe('Criterion Feedback', () => {
  const props = {
    orderNum: 1,
    config: 'config string',
    isGrading: true,
    value: 'some value',
    gradeStatus: 'ungraded',
    setValue: jest.fn().mockName('this.props.setValue'),
  };
  let el;
  beforeEach(() => {
    el = shallow(<CriterionFeedback {...props} />);
    el.instance().onChange = jest.fn().mockName('this.onChange');
  });
  describe('snapshot', () => {
    test('is grading', () => {
      expect(el.instance().render()).toMatchSnapshot();
    });

    test('is graded', () => {
      el.setProps({
        isGrading: false,
        gradeStatus: 'graded',
      });
      expect(el.instance().render()).toMatchSnapshot();
    });
  });

  describe('component', () => {
    test('is grading', () => {
      expect(el.isEmptyRender()).toEqual(false);
      expect(el.prop('value')).toEqual(props.value);
      expect(el.prop('disabled')).toEqual(false);
    });
    test('is graded', () => {
      el.setProps({
        isGrading: false,
        gradeStatus: 'graded',
      });
      expect(el.prop('value')).toEqual(props.value);
      expect(el.prop('disabled')).toEqual(true);
    });
    test('is ungraded and not grading', () => {
      el.setProps({
        isGrading: false,
      });
      expect(el.isEmptyRender()).toEqual(true);
    });
    test('set value called on change', () => {
      el = shallow(<CriterionFeedback {...props} />);
      el.instance().onChange({
        target: {
          value: 'some value',
        },
      });
      expect(props.setValue).toBeCalledTimes(1);
    });
  });

  describe('mapStateToProps', () => {
    const testState = { abitaryState: 'some data' };
    const additionalArgs = { orderNum: props.orderNum };
    let mapped;
    beforeEach(() => {
      mapped = mapStateToProps(testState, additionalArgs);
    });

    test('selectors.app.rubric.criterionFeedbackConfig', () => {
      expect(mapped.config).toEqual(
        selectors.app.rubric.criterionFeedbackConfig(testState, additionalArgs),
      );
    });

    test('selector.grading.selected.criterionFeedback', () => {
      expect(mapped.value).toEqual(
        selectors.grading.selected.criterionFeedback(testState, additionalArgs),
      );
    });

    test('selector.grading.selected.gradeStatus', () => {
      expect(mapped.gradeStatus).toEqual(
        selectors.grading.selected.gradeStatus(testState),
      );
    });
  });

  describe('mapDispatchToProps', () => {
    test('maps actions.grading.setCriterionFeedback to setValue prop', () => {
      expect(mapDispatchToProps.setValue).toEqual(
        actions.grading.setCriterionFeedback,
      );
    });
  });
});
