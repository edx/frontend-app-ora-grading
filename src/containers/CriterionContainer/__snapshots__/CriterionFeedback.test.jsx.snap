// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Criterion Feedback snapshot feedback is configured to disabled 1`] = `null`;

exports[`Criterion Feedback snapshot feedback is configured to optional 1`] = `
<Form.Group>
  <Form.Control
    as="textarea"
    className="criterion-feedback feedback-input"
    disabled={false}
    floatingLabel="Add comments (Optional)"
    onChange={[MockFunction this.onChange]}
    value="criterion value"
  />
</Form.Group>
`;

exports[`Criterion Feedback snapshot feedback is configured to required 1`] = `
<Form.Group>
  <Form.Control
    as="textarea"
    className="criterion-feedback feedback-input"
    disabled={false}
    floatingLabel="Add comments"
    onChange={[MockFunction this.onChange]}
    value="criterion value"
  />
</Form.Group>
`;

exports[`Criterion Feedback snapshot feedback value is invalid 1`] = `
<Form.Group>
  <Form.Control
    as="textarea"
    className="criterion-feedback feedback-input"
    disabled={false}
    floatingLabel="Add comments"
    onChange={[MockFunction this.onChange]}
    value="criterion value"
  />
  <Form.Control.Feedback
    className="feedback-error-msg"
    type="invalid"
  >
    The feedback is required
  </Form.Control.Feedback>
</Form.Group>
`;

exports[`Criterion Feedback snapshot is graded 1`] = `
<Form.Group>
  <Form.Control
    as="textarea"
    className="criterion-feedback feedback-input"
    disabled={true}
    floatingLabel="Comments"
    onChange={[MockFunction this.onChange]}
    value="criterion value"
  />
</Form.Group>
`;

exports[`Criterion Feedback snapshot is grading 1`] = `
<Form.Group>
  <Form.Control
    as="textarea"
    className="criterion-feedback feedback-input"
    disabled={false}
    floatingLabel="Add comments"
    onChange={[MockFunction this.onChange]}
    value="criterion value"
  />
</Form.Group>
`;
